lightning/src/ln/channelmanager.rs

```diff
@@ -2784,73 +2824,77 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
 	}
 
 	fn internal_channel_reestablish(&self, their_node_id: &PublicKey, msg: &msgs::ChannelReestablish) -> Result<(), MsgHandleErrInternal> {
-		let mut channel_state_lock = self.channel_state.lock().unwrap();
-		let channel_state = &mut *channel_state_lock;
+		let fail_htlcs;
+		{
+			let mut channel_state_lock = self.channel_state.lock().unwrap();
+			let channel_state = &mut *channel_state_lock;
 
-		match channel_state.by_id.entry(msg.channel_id) {
-			hash_map::Entry::Occupied(mut chan) => {
-				if chan.get().get_their_node_id() != *their_node_id {
-					return Err(MsgHandleErrInternal::send_err_msg_no_close("Got a message for a channel from the wrong node!".to_owned(), msg.channel_id));
-				}
-				let (funding_locked, revoke_and_ack, commitment_update, monitor_update_opt, mut order, shutdown) =
-					try_chan_entry!(self, chan.get_mut().channel_reestablish(msg, &self.logger), channel_state, chan);
-				if let Some(monitor_update) = monitor_update_opt {
-					if let Err(e) = self.monitor.update_monitor(chan.get().get_funding_txo().unwrap(), monitor_update) {
-						// channel_reestablish doesn't guarantee the order it returns is sensical
-						// for the messages it returns, but if we're setting what messages to
-						// re-transmit on monitor update success, we need to make sure it is sane.
-						if revoke_and_ack.is_none() {
-							order = RAACommitmentOrder::CommitmentFirst;
-						}
-						if commitment_update.is_none() {
-							order = RAACommitmentOrder::RevokeAndACKFirst;
+			match channel_state.by_id.entry(msg.channel_id) {
+				hash_map::Entry::Occupied(mut chan) => {
+					if chan.get().get_their_node_id() != *their_node_id {
+						return Err(MsgHandleErrInternal::send_err_msg_no_close("Got a message for a channel from the wrong node!".to_owned(), msg.channel_id));
+					}
+					let (funding_locked, revoke_and_ack, commitment_update, monitor_update_opt, htlcs_to_fail, mut order, shutdown) =
+						try_chan_entry!(self, chan.get_mut().channel_reestablish(msg, &self.logger), channel_state, chan);
+					if let Some(monitor_update) = monitor_update_opt {
+						if let Err(e) = self.monitor.update_monitor(chan.get().get_funding_txo().unwrap(), monitor_update) {
+							// channel_reestablish doesn't guarantee the order it returns is sensical
+							// for the messages it returns, but if we're setting what messages to
+							// re-transmit on monitor update success, we need to make sure it is sane.
+							if revoke_and_ack.is_none() {
+								order = RAACommitmentOrder::CommitmentFirst;
+							}
+							if commitment_update.is_none() {
+								order = RAACommitmentOrder::RevokeAndACKFirst;
+							}
+							return_monitor_err!(self, e, channel_state, chan, order, revoke_and_ack.is_some(), commitment_update.is_some());
```

Comment:
I think its a bug if we get htlcs_to_fail and then return early here without actually failing the HTLCs before returning (of course I don't know that we can get HTLCs to fail here to begin with, but its not ideal).

Reply:
That makes sense. However, my understanding from our slack conversation on Friday was that we can just assert that there are no htlcs to fail backwards (given that we drop all holding cell HTLC forwards on peer disconnect). Let me know if that isn't right.

Reply:
Yep! Can you add a more conservative assert?

```
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index cf2ccb9b..9e7b653e 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -2827,6 +2827,11 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
                                // update_adds should have been dropped on peer disconnect. If this changes in
                                // the future, corresponding changes will need to be made in the ChannelManager's
                                // reestablish logic, because the logic assumes there are no HTLCs to fail backwards.
+                               for htlc_update in self.holding_cell_htlc_updates.iter() {
+                                       if let &HTLCUpdateAwaitingACK::AddHTLC { .. } = htlc_update {
+                                               debug_assert!(false, "We don't handle some pending add-HTLC edge-cases properly on reconnect, and they shouldn't be there to begin with");
+                                       }
+                               }
                                match self.free_holding_cell_htlcs(logger) {
                                        Err(ChannelError::Close(msg)) => return Err(ChannelError::Close(msg)),
                                        Err(ChannelError::Ignore(_)) | Err(ChannelError::CloseDelayBroadcast(_)) => panic!("Got non-channel-failing result from free_holding_cell_htlcs"),
```